File: leetcode/763. Partition Labels.py
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        dic = [-1 for i in range(26)]
        for i, c in enumerate(S):
            dic[ord(c) - ord('a')] = i
        
        m = n = 0
        res = []
        for i in range(len(S)):
            c = S[i]
            n = max(dic[ord(c) - ord('a')], n)
            if i >= n:
                res.append(n - m + 1)
                m = n = i + 1
        return res
            

File: leetcode/763. Partition Labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string S) {
        vector<int> res;
        int dic[26];
        for(size_t i = 0; i < S.size(); ++i){
            dic[S[i] - 'a'] = i;
        }
        
        int m = 0;
        int n = 0;
        
        for(size_t i = 0; i < S.size(); ++i){
            n = max(n, dic[S[i] - 'a']);
            if(i >= n){
                res.push_back(n - m + 1);
                m = n = i + 1;
            }
        }
        
        return res;
    }
};
